homeassistant/components/poolsense/__init__.py

```diff
@@ -75,29 +78,51 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     return unload_ok
 
 
-async def get_coordinator(hass, entry):
-    """Get the data update coordinator."""
+class PoolSenseEntity(Entity):
+    """Implements a common class elements representing the PoolSense component."""
+
+    def __init__(self, coordinator, email, password, info_type):
+        """Initialize poolsense sensor."""
+        self._email = email
+        self._password = password
+        self._unique_id = f"{email}-{info_type}"
+        self.coordinator = coordinator
+        self.info_type = info_type
+
+    @property
+    def unique_id(self):
+        """Return a unique id."""
+        return self._unique_id
+
+    @property
+    def available(self):
+        """Return if sensor is available."""
+        return self.coordinator.last_update_success
+
 
-    async def async_get_data():
-        _LOGGER.info("Run query to server")
-        poolsense = PoolSense()
-        return_data = {}
+class PoolSenseDataUpdateCoordinator(DataUpdateCoordinator):
+    """Define an object to hold PoolSense data."""
+
+    def __init__(self, hass, entry):
+        """Initialize."""
+        self.poolsense = PoolSense()
+        self.hass = hass
+        self.entry = entry
+
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=timedelta(hours=1))
+
+    async def _async_update_data(self):
+        """Update data via library."""
+        data = {}
         with async_timeout.timeout(10):
             try:
-                return_data = await poolsense.get_poolsense_data(
-                    aiohttp_client.async_get_clientsession(hass),
-                    entry.data[CONF_EMAIL],
-                    entry.data[CONF_PASSWORD],
+                data = await self.poolsense.get_poolsense_data(
```

Comment:
this seems a bit anti-pattern to me and i think Martin was pointing it out before but I think PoolSense class should accept the session and login info and internally share it when needed to make these calls.

Reply:
Thanks. I have changed the structure to let the Poolsense class deal with the credentials as suggested.